[Spaces] - Space aware privileges UI

legrego opened the pull request:
This PR includes enhancements to the Role Management screen to allow users to specify Kibana Privileges on a per-space level.

This PR does not include changes to *enforce* space-aware privileges; this only includes the UI/API changes necessary to support editing a role's privileges.

## Granting access to specific spaces
![image](https://user-images.githubusercontent.com/3493255/43021742-d2e7961a-8c32-11e8-8a5a-d60f3d9059e5.png)

## Granting full access to all spaces
![image](https://user-images.githubusercontent.com/3493255/43021754-e187a232-8c32-11e8-9e88-bbf7521d8e2a.png)

## Granting read access to all spaces
![image](https://user-images.githubusercontent.com/3493255/43021779-f11dc19a-8c32-11e8-90f6-5706725ccc2e.png)

## Granting read access to all spaces, but full access to the Marketing Space
![image](https://user-images.githubusercontent.com/3493255/43021813-09cc6156-8c33-11e8-8dc9-9e25edf47b5f.png)

## Viewing existing space permissions alongside current edits
![image](https://user-images.githubusercontent.com/3493255/43021892-51dcc67a-8c33-11e8-9fb1-eb1f96af2503.png)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7459/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7470/)

legrego commented:
@kobelb -- this PR isn't ready for review yet, but my work here exposes a potential pain-point that I want to discuss:

From an API perspective, having to specify the auto-generated space id when assigning privileges to spaces is going to be cumbersome and not very user-friendly.

I know at one point we weren't keen on the idea, but I'm thinking of "rebranding" the `URL Context` to become the `Public Space ID`. From a URL perspective, things would remain unchanged, but we would use this public space id as part of the actual document id when creating the saved object that represents the space.

For example, a space that used to be stored like this:
```javascript
{
	"_id": "space:abc-123-gef-45-334....",
	"type": "space",
	"space": {
		"name": "My First Space",
		"urlContext": "my-first-space",
		"description": "This is my first space"
	}
}
```

Would end up getting stored like this instead:
```javascript
{
	"_id": "space:my-first-space",
	"type": "space",
	"space": {
		"name": "My First Space",
		"description": "This is my first space"
	}
}
```

With this approach, the Space ID becomes human-readable, human-selectable, and is prominently displayed in the UI.

API benefits aside, this would also:
1) Provide a guarantee that we won't have duplicate URL Contexts
2) Eliminate a round-trip on certain requests


The only potential downside I can think of is import/export, but I don't think we are planning on exposing these objects via the import/export UI/API, so I think it's a non-issue.

What are your thoughts?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8025/)

legrego commented:
Hey @cchaos are you available for a design review? I based this off of your mockups in [this comment](https://github.com/elastic/kibana/issues/19449#issuecomment-397373567), so hopefully, there won't be a ton of work for you to do here 😄 

You don't need to use a special branch of ES to test this anymore, now that we've merged 6.3 to master -- you should be fine to use either master or snapshot, however you usually start ES

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8312/)

cchaos commented:
@legrego I created a design PR for you: https://github.com/legrego/kibana/pull/8

Please also work with @gchaps on copy.

A couple things to think about:

1. Adding search and bulk actions to the table. This would help mitigate the issue of creating in bulk vs saving and displaying individually.
2. What happens when they change from "none" to "read" base privilege and there were already space-specific read privileges. Right now it's kind of ok that with a "read" base privilege, you have the custom options of read or all. However, down the line when more privileges are created, there may need to be a conversion of sorts.

legrego commented:
Hey @gchaps,

The Role Management screen in this PR has been updated to allow users to set Kibana privileges for specific spaces. Additionally, they can set a "minimum privilege" that applies to all spaces. For example, they can specify that a particular role can have "read only" access to all spaces, but maybe "write" access to the Marketing space.

Depending on the minimum privilege, we will show a message explaining the options/limitations for setting per-space privileges. Would you be able to work with us on the copy for these messages? I'm happy to walk through any of this in more detail if that would help:

### Minimum Privilege set to "all":
> This role always grants full access to all spaces. To customize privileges for individual spaces, you must create a new role.
![image](https://user-images.githubusercontent.com/3493255/43519563-ef4c8fde-955d-11e8-8732-19c4eb2718b1.png)


### Minimum Privilege set to "read":
> Lowest possible privilege is "read"
![image](https://user-images.githubusercontent.com/3493255/43519604-1cd2b8fc-955e-11e8-9f0a-8470d58fe408.png)


### Reserved "kibana_user" role:
> This role always grants full access to all spaces. To customize privileges for individual spaces, you must create a new role.
![image](https://user-images.githubusercontent.com/3493255/43519656-4e821d52-955e-11e8-9561-b09a7cb36b3d.png)


### Reserved "kibana_dashboard_only_user" role:
> This role always grants read access to all spaces. To customize privileges for individual spaces, you must create a new role.
![image](https://user-images.githubusercontent.com/3493255/43519685-709b3d60-955e-11e8-81e1-7988904f2cdf.png)


### Reserved role without any access to Kibana:
> This role never grants access to any spaces within Kibana. To customize privileges for individual spaces, you must create a new role.
![image](https://user-images.githubusercontent.com/3493255/43519727-8f2fca16-955e-11e8-861c-cba59b6bfd24.png)

legrego commented:
@cchaos 

> Adding search and bulk actions to the table. This would help mitigate the issue of creating in bulk vs saving and displaying individually.

Are you thinking of adding the bulk actions form as a row in the table, or are you thinking of having a button they click which adds the individual results of the form to the table prior to saving?


> What happens when they change from "none" to "read" base privilege and there were already space-specific read privileges. Right now it's kind of ok that with a "read" base privilege, you have the custom options of read or all. However, down the line when more privileges are created, there may need to be a conversion of sorts.

Absolutely -- I wasn't going to spend a whole of time on that scenario just yet, but I agree that we'll likely need something more robust once more privileges are created.

gchaps commented:
@legrego Thanks for the screenshots of the error messages.  Will review the text.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/8566/)

AlonaNadler commented:
I checked this PR and had a few suggestions

![image](https://user-images.githubusercontent.com/32550316/43603639-5cd6320e-9648-11e8-9be3-ef36e06b015d.png)
- When minimal priviliges is `all` can we show a list of all spaces with the ` all` privilige for each space, I think it will be more explicit and easier to understand which Spaces this role will have access to.
- When the Lowest possible privilege is 'read' can we still show `none`  in the drop-down just disabled.
- The index privliges now has a button while `run as priviliges` and `cluster priviliges`  are presented in the page, why is it better approach ? while `run as privileges` and `cluster privileges` is not used with every Role, Index privileges is almost a must, I'm not sure I understand the reasoning why part of the page is editable and part requires to click on a button 
- How about adding a link from  create a spaces page to the roles page as a call to action, otherwise the flow is going back to management and moving to roles,

legrego commented:
@cchaos 
>  I'm talking about utilizing the search and selection options for a table, to add something similar to .... Where two buttons appear, "Delete" and "Edit", the Edit may be a popover with the privilege select and an "Apply to selected" button. That way they can edit more than one at a time and possibly delete all at once without needing to click each individually.

I understand now -- thanks for clarifying! I'll see if I can add this in without a ton of additional effort. If it becomes too big of a change, then I may postpone.

@AlonaNadler 
> When minimal priviliges is all can we show a list of all spaces with the all privilige for each space, I think it will be more explicit and easier to understand which Spaces this role will have access to.

We _could_ do this, but I'm worried this would give the impression that those are the only impacted spaces, when that's not necessarily the case. Any space created after the role is defined is also impacted by the minimum privilege. This list could get rather long too, for users with a lot of spaces.

> When the Lowest possible privilege is 'read' can we still show none in the drop-down just disabled.

Do we have a precident either way in Kibana for this vs hiding the option? I'm not necessarily opposed, but I'd like to be consistent.

> The index privliges now has a button while run as priviliges and cluster priviliges are presented in the page, why is it better approach ? while run as privileges and cluster privileges is not used with every Role, Index privileges is almost a must, I'm not sure I understand the reasoning why part of the page is editable and part requires to click on a button

That's an interesting point. The Index Privileges has the button because you are able to add multiple, whereas the other privileges have a fixed set of input fields. I think that setting index privileges is less necessary than before though. I assume that roles are created with index privileges a majority of the time, but there are also cases where you don't need them defined for a particular role. What if we added in a blank form to get them started, but keep the button if they wanted to add more than one set of index privileges?

> How about adding a link from create a spaces page to the roles page as a call to action, otherwise the flow is going back to management and moving to roles,

We want the Create Spaces page to be easier to use when security is enabled, but we are intentionally keeping these types of features out of Phase 1. This is certainly a candidate for a future release though!

cchaos commented:
Here are some mocks about how we can handle a summary view. Keep in mind some parts of the form are out of date. The main idea is the link below the callout that opens a flyout that has a table listing ALL spaces and their respective privileges.

![kibana options copy](https://user-images.githubusercontent.com/549577/43910882-03c5e69e-9bcc-11e8-9842-b66da0f30152.png)

![kibana options copy 2](https://user-images.githubusercontent.com/549577/43910888-061f5dbc-9bcc-11e8-97f0-3e476cddad9b.png)

cc @AlonaNadler @kobelb

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/97/)

legrego commented:
I added my first-pass at implementing the impacted spaces flyout. The "Manage Spaces" button is stubbed out right now (not functional) because I plan on implementing that within the Spaces plugin so that I can use it in multiple places.

cchaos commented on x-pack/plugins/security/public/views/management/edit_role/components/privileges/kibana/privilege_callout_warning.js:
```diff
@@ -3,69 +3,159 @@
+                To customize privileges for individual spaces, you must create a new role.
+              </p>
+            </EuiCallOut>
+            {this.getShowImpactedSpacesLink()}
```
Can this show no matter what the base privilege is? So you can always get a full summary list?

cchaos replied:
And put it in line with the "Add space privilege" button.

cchaos commented on x-pack/plugins/security/public/views/management/edit_role/components/privileges/kibana/privilege_callout_warning.js:
```diff
@@ -3,69 +3,159 @@
+    }, {});
+
+    return (
+      <EuiFlyout onClose={this.toggleShowImpactedSpaces} aria-labelledby="showImpactedSpacesTitle">
```
Use `size="s"` for this flyout

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/100/)

cchaos commented on x-pack/plugins/security/public/views/management/edit_role/components/privileges/kibana/privilege_space_table.js:
```diff
@@ -0,0 +1,149 @@
+      field: 'privilege',
+      name: 'Privilege',
+      render: (privilege, record) => {
+        if (this.props.readonly) {
```
Can you reduce the width of this column if it is read only? That will give more space to the space name.

cchaos commented on x-pack/plugins/security/public/views/management/edit_role/components/privileges/kibana/privilege_callout_warning.js:
```diff
@@ -0,0 +1,161 @@
+        <EuiFlyoutBody>
+          <PrivilegeSpaceTable readonly={true} role={role} spaces={spaces} spacePrivileges={spacePrivileges} />
+        </EuiFlyoutBody>
+        <EuiFlyoutFooter className="privilegeCallout--footer">
```
I would just hide the entire footer if the manage button is not available.

legrego replied:
yep, once we have that mechanism in place, I'll go back and update this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/108/)

cchaos commented on x-pack/plugins/security/public/views/management/edit_role/components/privileges/kibana/space_aware_privilege_form.js:
```diff
@@ -150,24 +158,40 @@ export class SpaceAwarePrivilegeForm extends Component {
-                {availableSpaces.length > 0 &&
+            {this.getSpaceForms(basePrivilege)}
+
+            <EuiFlexGroup responsive={false} alignItems={'flexEnd'}>
```
I would leave `responsive` as `true` (or just remove this prop) and change `alignItems` to 'baseline' so the text in the button and the text link line up.

legrego replied:
good call, thanks!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/118/)

cchaos commented:
Looks good! I'm just seeing a console error about `controlled or uncontrolled` for the privilege select component.

legrego commented:
> Looks good! I'm just seeing a console error about controlled or uncontrolled for the privilege select component.

Thanks for your help @cchaos! Yeah, I'm still trying to track that one down

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/121/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/138/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/147/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/269/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/389/)

legrego commented:
@kobelb this is ready for review -- I still have tests to add, but the rest of the code is fine for you to start looking at. The screenshots above are rather outdated at this point, but I can update for posterity before we merge.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/391/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/392/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/456/)

kobelb commented on x-pack/plugins/security/public/views/management/edit_role/lib/get_application_privileges.js:
```diff
@@ -3,23 +3,15 @@
-    // we don't want to display privileges that aren't in our expected list of privileges
-    if (assignedPrivilege in viewModel) {
-      viewModel[assignedPrivilege] = true;
+export function getKibanaPrivilegesViewModel(applicationPrivileges, roleKibanaPrivileges) {
```
Yay! It seems like we don't really need this anymore now that we aren't rendering the "checkbox list" you think we can remove it?

kobelb commented on x-pack/plugins/spaces/server/routes/api/v1/spaces.js:
```diff
@@ -143,8 +144,10 @@ export function initSpacesApi(server) {
 
       let result;
 
+      let existingSpace;
```
This doesn't seem to need to be a `let`, am I missing something?

kobelb commented on x-pack/plugins/security/public/views/management/edit_role/components/privileges/kibana/impacted_spaces_flyout.js:
```diff
@@ -0,0 +1,100 @@
+    const { assignedPrivileges } = getKibanaPrivilegesViewModel(kibanaAppPrivileges, role.kibana);
+
+    const allSpacePrivileges = spaces.reduce((acc, space) => {
+      const spacePrivilege = assignedPrivileges.space[space.id] ? assignedPrivileges.space[space.id][0] : basePrivilege;
```
If the user adds a Space privilege of "read" and then sets the global privilege of "all", we still show the read privilege in the flyout.

legrego commented on x-pack/plugins/security/index.js:
```diff
@@ -27,7 +27,7 @@ export const security = (kibana) => new kibana.Plugin({
   configPrefix: 'xpack.security',
   publicDir: resolve(__dirname, 'public'),
-  require: ['kibana', 'elasticsearch', 'xpack_main'],
+  require: ['kibana', 'elasticsearch', 'xpack_main', 'spaces'],
```
Shouldn't take more than a day -- I have old code that I can resurrect from history as a starting point

legrego commented on x-pack/plugins/security/public/views/management/edit_role/components/privileges/kibana/space_aware_privilege_form.js:
```diff
@@ -0,0 +1,316 @@
+            If a space is not customized, its permissions will default to the minimum privilege specified above.
+          </p>
+          {basePrivilege !== 'all' && this.props.editable && (
+            <p>You can bulk-create space privileges though they will be saved individually upon saving the role.</p>
```
This is just poor wording that we'll get Gail's help on. The save operation is atomic, this is referring to how the privileges will be represented in this UI if you save/reload the screen

legrego replied:
This is just poor wording that we'll get Gail's help on. The save operation is atomic, this is referring to how the privileges will be represented in this UI if you save/reload the screen

kobelb replied:
I was attempting to refer to the behavior that we see when assigning the privilege to multiple spaces, if we assign read to two spaces using a single "line", when we go to save it and then re-open it it's been split:

<img width="883" alt="screen shot 2018-08-21 at 7 22 08 am" src="https://user-images.githubusercontent.com/627123/44400297-acb2d780-a518-11e8-9897-bfd6cf14c2f1.png">

I'm not sure if this is what users will want or not... did design have any input on this behavior?

legrego commented on x-pack/plugins/security/public/views/management/edit_role/components/privileges/kibana/space_selector.test.js:
```diff
@@ -0,0 +1,15 @@
+
+describe('SpaceSelector', () => {
+  it('renders without crashing', () => {
+    expect(mount(<SpaceSelector spaces={[]} selectedSpaceIds={[]} onChange={jest.fn()} />)).toMatchSnapshot();
```
Yep good catch. Will change!

legrego commented:
@kobelb I think I addressed all of your comments, if you'd like to re-review.
I also added missing validations to the space permission ui, so it's more user-friendly when trying to save incomplete settings

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/873/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/903/)

legrego commented:
Merged into our feature branch since we have a LGTM, but I'm aware we have copy edits to make yet. I'll open a separate PR to address those once available.
